AWS PCA Issuer

anshrma opened the pull request:
### What does this PR do?

This PR adds a new module for helm installation of AWS PCA issuer, which is an extension of cert-manager for certificate management using AWS ACM. 


### Motivation

Working with a large Partner in API Gateway space (another PR coming next), which requires this


### More

- [ Y] Yes, I have tested the PR using my local account setup  (Provide any test evidence report under Additional Notes)
- [ NA] Yes, I have added a new example under [examples](https://github.com/aws-samples/aws-eks-accelerator-for-terraform/tree/main/examples) to support my PR
- [NA] Yes, I have created another PR for add-ons under [add-ons](https://github.com/aws-samples/ssp-eks-add-ons) repo (if applicable)
- [ Y] Yes, I have updated the [docs](https://github.com/aws-samples/aws-eks-accelerator-for-terraform/tree/main/docs) for this feature
- [ ] Yes, I ran `pre-commit run -a` with this PR


**Note**: Not all the PRs required examples and docs except a new pattern or add-on added.

### For Moderators
- [ ] E2E Test successfully complete before merge?

### Additional Notes

<!-- Anything else we should know when reviewing? -->

vara-bonthu commented on modules/kubernetes-addons/aws-pca-issuer/main.tf:
```diff
@@ -0,0 +1,23 @@
+module "cert_manager" {
+  source            = "./../cert-manager"
+  helm_config       = var.cert_manager_helm_config
+  manage_via_gitops = var.manage_via_gitops
+  addon_context     = var.addon_context
+}
```
cert manager should be deployed from the consumer example . This add-on should only install AWS PCA Issuer

anshrma replied:
hi @vara-bonthu - Any pointers on how customers add dependency between two modules ? 

For example, aws-pca-issuer depends on cert-manager. Customer enable `cert-manager` by `enable_cert_manager                 = true` and pca issuer by `enable_aws_pca_issuer               = true` . Which is why, i added the dependency aws-pca-issuer itself, which requires `cert-manager` as a pre-req.

askulkarni2 replied:
We need to design a proper dependency mechanism between addons. @anshrma at this point please document the dependency and update your usage example to show `enable_cert_manager = true` is needed.

vara-bonthu commented on modules/kubernetes-addons/aws-pca-issuer/variables.tf:
```diff
@@ -0,0 +1,46 @@
+  description = "Input configuration for the addon"
+}
+
+#-----------CERT MANAGER-------------
+variable "enable_cert_manager" {
+  type        = bool
+  default     = true
+  description = "Enable Cert Manager add-on"
+}
+
+variable "cert_manager_helm_config" {
+  type        = any
+  description = "Cert Manager Helm Chart config"
+  default     = {}
+}
```
remove these. see the above comment

vara-bonthu commented on modules/kubernetes-addons/main.tf:
```diff
@@ -122,6 +122,14 @@ module "cert_manager" {
+  source            = "./aws-pca-issuer"
+  helm_config       = var.aws_pca_issuer_helm_config
+  manage_via_gitops = var.argocd_manage_add_ons
+  addon_context     = local.addon_context
```
need to pass `irsa_iam_permissions_boundary`

askulkarni2 replied:
role_path and permission_boundary is moving to `addon_context` in [PR](https://github.com/aws-samples/aws-eks-accelerator-for-terraform/pull/351)

vara-bonthu reviewed (changes requested):
Thanks @anshrma for the PR. This will be a great add-on. Would you mind creating a pattern to show with example on how to use this Add-on?

Update docs page as well here `https://github.com/aws-samples/aws-eks-accelerator-for-terraform/tree/main/docs/add-ons`

askulkarni2 commented on modules/kubernetes-addons/aws-pca-issuer/locals.tf:
```diff
@@ -0,0 +1,47 @@
+    create_kubernetes_service_account = true
+    kubernetes_service_account        = local.service_account_name
+    irsa_iam_policies                 = [aws_iam_policy.aws_pca_issuer.arn]
+    irsa_iam_permissions_boundary     = var.irsa_iam_permissions_boundary
```
This is going to move to `addon_context`. Please keep an eye on this [PR](https://github.com/aws-samples/aws-eks-accelerator-for-terraform/pull/351).

askulkarni2 commented on modules/kubernetes-addons/aws-pca-issuer/variables.tf:
```diff
@@ -0,0 +1,46 @@
+  description = "Determines if the add-on should be managed via GitOps."
+}
+
+variable "irsa_iam_permissions_boundary" {
+  type        = string
+  default     = ""
+  description = "IAM Policy ARN for IRSA IAM role permissions boundary"
+}
```
Needs to be updated if [PR](https://github.com/aws-samples/aws-eks-accelerator-for-terraform/pull/351) is merged before this one.

askulkarni2 commented on modules/kubernetes-addons/aws-pca-issuer/variables.tf:
```diff
@@ -0,0 +1,46 @@
+  description = "IAM Policy ARN for IRSA IAM role permissions boundary"
+}
+
+variable "addon_context" {
```
Same as above.

askulkarni2 commented on modules/kubernetes-addons/main.tf:
```diff
@@ -122,6 +122,14 @@ module "cert_manager" {
 }
 
+module "aws_pca_issuer" {
+  count             = var.enable_aws_pca_issuer ? 1 : 0
```
Update to `enable_aws_privateca_issuer`

askulkarni2 commented on modules/kubernetes-addons/main.tf:
```diff
@@ -122,6 +122,14 @@ module "cert_manager" {
 
+module "aws_pca_issuer" {
+  count             = var.enable_aws_pca_issuer ? 1 : 0
+  source            = "./aws-pca-issuer"
```
`aws-privateca-issuer`

askulkarni2 commented on modules/kubernetes-addons/variables.tf:
```diff
@@ -369,6 +369,19 @@ variable "cert_manager_helm_config" {
 }
 
+#-----------AWS PCA ISSUER-------------
+variable "enable_aws_pca_issuer" {
```
`enable_aws_privateca_issuer`

askulkarni2 commented on modules/kubernetes-addons/variables.tf:
```diff
@@ -369,6 +369,19 @@ variable "cert_manager_helm_config" {
+  description = "Enable PCA Issuer"
+}
+
+variable "aws_pca_issuer_helm_config" {
```
`aws_privateca_issuer_helm_config`

anshrma replied:
addressed

askulkarni2 reviewed (commented):
@anshrma thanks so much for the PR. Its looking good, some minor changes are needed as per our convention. Also please address @vara-bonthu's comments as well.

anshrma commented:
@vara-bonthu / @askulkarni2 - Thanks for review. Addressed all the feedback with updated code. Pls let me know, if anything else. Once  this gets merged, next i will send another PR for a partner using these add-ons

anshrma commented on modules/kubernetes-addons/cert-manager/locals.tf:
```diff
@@ -21,22 +21,13 @@ locals {
     var.helm_config
   )
 
-  set_values = [
```
@askulkarni2 / @vara-bonthu - As aws-privateca-issuer depends on cert manager, i received the following error in cert-manager pod, which was resulting in certificate requests getting failed to be mounted as secrets.

```
E0322 01:59:36.653678       1 controller.go:163] cert-manager/certificaterequests-approver "msg"="re-queuing item due to error processing" "error"="admission webhook \"webhook.cert-manager.io\" denied the request: status.conditions: Forbidden: user \"system:serviceaccount:cert-manager:cert-manager-sa\" does not have permissions to set approved/denied conditions for issuer {aws001-preprod-dev-eks AWSPCAClusterIssuer awspca.cert-manager.io}" "key"="default/example-nftjp" 
```

Helm chart installation of cert-manager would pass and pods kept running, but when used by aws-privateca-issuer or any other Issuer, it would log errors.

On digging further, i realized that these two parameters were culprit and was incorrectly set, resulting in cert-manager RBAC to not work as it just created a SA without any permissions as what the helm chat typically would do. IRSA is not required here as its the PCA pod , which uses IRSA. [cert-manager](https://cert-manager.io/docs/installation/helm/) helm chart guide also has no such requirements.

askulkarni2 replied:
Can we confirm that we do not need IRSA for any other use cases? If not then `irsa_config` can be removed altogether and just create the namespace instead such as [here](https://github.com/aws-samples/aws-eks-accelerator-for-terraform/blob/dbcb4fe9e5da10032f30dea2ed3f3946808d7c0e/modules/kubernetes-addons/ingress-nginx/main.tf#L11).

vara-bonthu replied:
@anshrma Its because Private CA add-on is expecting the  cert-manager **service account name** as `cert-manager`

vara-bonthu replied:
@askulkarni2 Yes, we need IRSA for cert-manager

vara-bonthu replied:
with a policy like
```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Sid": "AllowRoute53",
            "Effect": "Allow",
            "Action": [
                "route53:ChangeResourceRecordSets",
                "route53:ListResourceRecordSets"
            ],
            "Resource": [
                "arn:aws:route53:::hostedzone/${hosted_zone_id}"
            ]
        },
        {
            "Sid": "AllowRoute53-GetChange",
            "Effect": "Allow",
            "Action": [
                "route53:GetChange"
            ],
            "Resource": [
                "arn:aws:route53:::change/*"
            ]
        },
        {
            "Sid": "AllowRoute53-ListHostedZonesByName",
            "Effect": "Allow",
            "Action": [
                "route53:ListHostedZonesByName"
            ],
            "Resource": [
                "*"
            ]
        }
    ]
}
```

askulkarni2 commented on modules/kubernetes-addons/main.tf:
```diff
@@ -122,6 +122,15 @@ module "cert_manager" {
+  source            = "./aws-privateca-issuer"
+  helm_config       = var.aws_privateca_issuer_helm_config
+  manage_via_gitops = var.argocd_manage_add_ons
+  irsa_iam_permissions_boundary = var.aws_privateca_issuer_irsa_permissions_boundary
```
Can be removed from here and passed via `addon_context` as mentioned in the above comments.

askulkarni2 commented on modules/kubernetes-addons/variables.tf:
```diff
@@ -369,6 +369,25 @@ variable "cert_manager_helm_config" {
+  default     = {}
+}
+
+variable "aws_privateca_issuer_irsa_permissions_boundary" {
+  type        = string
+  description = "IAM Policy ARN for IRSA IAM role permissions boundary"
+  default     = ""
+}
+
```
Remove

askulkarni2 commented on modules/kubernetes-addons/aws-privateca-issuer/variables.tf:
```diff
@@ -0,0 +1,33 @@
+    tags                           = map(string)
+  })
+  description = "Input configuration for the addon"
+}
```
Add `aws_privateca_acmca_arn` as input.

askulkarni2 commented on modules/kubernetes-addons/main.tf:
```diff
@@ -122,6 +122,15 @@ module "cert_manager" {
+  count             = var.enable_aws_privateca_issuer ? 1 : 0
+  source            = "./aws-privateca-issuer"
+  helm_config       = var.aws_privateca_issuer_helm_config
+  manage_via_gitops = var.argocd_manage_add_ons
```
Pass `aws_privateca_acmca_arn`

askulkarni2 commented on modules/kubernetes-addons/variables.tf:
```diff
@@ -369,6 +369,25 @@ variable "cert_manager_helm_config" {
+  description = "PCA Issuer Helm Chart config"
+  default     = {}
+}
+
```
Add `aws_privateca_acmca_arn` as input

askulkarni2 reviewed (commented):
@anshrma some nitpicky comments, but PR is looking really good. 

One question on whether we should remove all IRSA for `cert-manager`. I don't have enough knowledge there, will wait on a few confirmations there from @vara-bonthu, @kcaws.

anshrma commented:
@askulkarni2 / @vara-bonthu - Addressed all the review feedback.

askulkarni2 reviewed (approved):
@anshrma thanks for the updates. LGTM!

vara-bonthu commented on examples/tls-with-aws-pca-issuer/main.tf:
```diff
@@ -0,0 +1,233 @@
+  environment        = "preprod" # Environment area eg., preprod or prod
+  zone               = "dev"     # Environment with in one sub_tenant or business unit
+  kubernetes_version = "1.21"
+
```
you might need this variable here   
`azs          = slice(data.aws_availability_zones.available.names, 0, 3)`

vara-bonthu commented on examples/tls-with-aws-pca-issuer/main.tf:
```diff
@@ -0,0 +1,233 @@
+
+  name = local.vpc_name
+  cidr = local.vpc_cidr
+  azs  = data.aws_availability_zones.available.names
```
`azs = local.azs`

vara-bonthu commented on examples/tls-with-aws-pca-issuer/main.tf:
```diff
@@ -0,0 +1,233 @@
+  cidr = local.vpc_cidr
+  azs  = data.aws_availability_zones.available.names
+
+  public_subnets  = [for k, v in data.aws_availability_zones.available.names : cidrsubnet(local.vpc_cidr, 8, k)]
+  private_subnets = [for k, v in data.aws_availability_zones.available.names : cidrsubnet(local.vpc_cidr, 8, k + 10)]
```
```hcl
  public_subnets  = [for k, v in local.azs : cidrsubnet(local.vpc_cidr, 8, k)]
  private_subnets = [for k, v in local.azs : cidrsubnet(local.vpc_cidr, 8, k + 10)]
```

vara-bonthu commented on examples/tls-with-aws-pca-issuer/main.tf:
```diff
@@ -0,0 +1,233 @@
+    }
+  }
+
+  depends_on = [module.aws_vpc]
```
I believe TF plan doesn't work with this `depends_on = [module.aws_vpc]` option with the latest tag and it requires to use `-target`

vara-bonthu commented on examples/tls-with-aws-pca-issuer/main.tf:
```diff
@@ -0,0 +1,233 @@
+  certificate       = aws_acmpca_certificate.example.certificate
+  certificate_chain = aws_acmpca_certificate.example.certificate_chain
+}
+
```
It would be nice to add what this resource does. e.g., 
```
#-------------------------------
#  This resource issues a certificate using AWS Certificate Manager Private Certificate Authority (ACM PCA)
#-------------------------------
```

vara-bonthu commented on examples/tls-with-aws-pca-issuer/main.tf:
```diff
@@ -0,0 +1,233 @@
+  }
+}
+
+resource "aws_acmpca_certificate_authority" "example" {
```
Note required. Same as above comment

vara-bonthu commented on modules/kubernetes-addons/aws-privateca-issuer/locals.tf:
```diff
@@ -0,0 +1,46 @@
+    kubernetes_namespace              = local.name
+    create_kubernetes_service_account = true
+    kubernetes_service_account        = local.service_account_name
+    irsa_iam_policies                 = [aws_iam_policy.aws_privateca_issuer.arn]
```
We need to provide options to add additional IAM policies to this IAM role
```hcl
irsa_iam_policies                 = concat([aws_iam_policy.aws_privateca_issuer.arn], var.irsa_policies)
```

and add the following to `varriables.tf`

```hcl
variable "irsa_policies" {
  type        = list(string)
  description = "Additional IAM policies for a IAM role for service accounts"
  default     = []
}
```

vara-bonthu commented on modules/kubernetes-addons/aws-privateca-issuer/variables.tf:
```diff
@@ -0,0 +1,33 @@
+  default     = {}
+  description = "AWS PCA Issuer Helm Config"
+}
+
```
Add this one
```hcl
variable "irsa_policies" {
  type        = list(string)
  description = "Additional IAM policies for a IAM role for service accounts"
  default     = []
}
```

vara-bonthu commented on modules/kubernetes-addons/cert-manager/locals.tf:
```diff
@@ -20,22 +20,13 @@ locals {
   )
 
-  set_values = [
-    {
-      name  = "serviceAccount.name"
-      value = local.service_account_name
-    },
-    {
-      name  = "serviceAccount.create"
-      value = false
-    }
-  ]
```
First change the following 

from

```hcl
service_account_name = "${local.name}-sa"
```

to

```hcl
service_account_name = "cert-manager"
```

this block should stay with  `local.service_account_name`

Also, add a note here saying AWS PrivateCA is expecting the service account name as `cert-manager`. refer to docs if you can

vara-bonthu commented on modules/kubernetes-addons/cert-manager/locals.tf:
```diff
@@ -20,22 +20,13 @@ locals {
 
   irsa_config = {
     kubernetes_namespace              = local.helm_config["namespace"]
     kubernetes_service_account        = local.service_account_name
```
```hcl
kubernetes_service_account        = "cert-manager"
```

vara-bonthu commented on modules/kubernetes-addons/cert-manager/locals.tf:
```diff
@@ -20,22 +20,13 @@ locals {
     kubernetes_namespace              = local.helm_config["namespace"]
     kubernetes_service_account        = local.service_account_name
     create_kubernetes_namespace       = true
-    create_kubernetes_service_account = true
```
Keep this as true. This will create SA as `cert-manager`

vara-bonthu commented on modules/kubernetes-addons/main.tf:
```diff
@@ -102,6 +102,15 @@ module "cert_manager" {
+  helm_config               = var.aws_privateca_issuer_helm_config
+  manage_via_gitops         = var.argocd_manage_add_ons
+  addon_context             = local.addon_context
+  aws_privateca_acmca_arn   = var.aws_privateca_acmca_arn
```
you also need

```hcl
  irsa_policies            = var.aws_privateca_issuer_irsa_policies
```

vara-bonthu commented on modules/kubernetes-addons/variables.tf:
```diff
@@ -389,6 +389,25 @@ variable "cert_manager_helm_config" {
+  description = "PCA Issuer Helm Chart config"
+  default     = {}
+}
+
```
Add this variable `aws_privateca_issuer_irsa_policies`

vara-bonthu reviewed (changes requested):
Good work on this PR @anshrma. 

Please see comments and it requires few changes.

vara-bonthu commented:
> 

Yes, I have provided the IRSA policies above

anshrma commented on examples/tls-with-aws-pca-issuer/README.md:
````diff
@@ -0,0 +1,117 @@
+The following command destroys the resources created by `terraform apply`
+
+```shell script
+cd examples/eks-cluster-with-new-vpc
````
I believe, you are looking at old version, https://github.com/aws-samples/aws-eks-accelerator-for-terraform/pull/359/files#diff-b7042f890e5ccf8d973d957f8059ff468fa9db7bd6a1697e56c2767ad102419bR29 already has this change

askulkarni2 commented:
@anshrma have you had a chance to review @vara-bonthu's feedback?

anshrma commented:
> @anshrma have you had a chance to review @vara-bonthu's feedback?

Hi @askulkarni2 - Will pick this up tomorrow. Back today after PTO. Thanks

anshrma commented:
Thanks for all the feedback @vara-bonthu . Done with all the changes suggested.

anshrma commented:
hi @vara-bonthu / @askulkarni2 - Pls review this when you get a chance. All the feedback comments have been addressed.

vara-bonthu reviewed (approved):
Looks good 👍🏼  Needs to run E2E

askulkarni2 commented:
@anshrma I am running our e2e tests and I see some `terraform fmt` errors. See https://github.com/aws-ia/terraform-aws-eks-blueprints/runs/6201070488?check_suite_focus=true. Can you please fix them?

FYI.. I am also running terratest here https://github.com/aws-ia/terraform-aws-eks-blueprints/runs/6201005974?check_suite_focus=true.

anshrma commented:
> @anshrma I am running our e2e tests and I see some `terraform fmt` errors. See https://github.com/aws-ia/terraform-aws-eks-blueprints/runs/6201070488?check_suite_focus=true. Can you please fix them?
> 
> FYI.. I am also running terratest here https://github.com/aws-ia/terraform-aws-eks-blueprints/runs/6201005974?check_suite_focus=true.

hi @askulkarni2 - fixed the terraform fmt errors
